### PR TITLE
Codebase Cleanup, Badger Query Optimization, PostgreSQL Index Creation

### PR DESCRIPTION
**Pull request by bonedaddy**
* Slightly optimize badger query function
* Add testutils package, closing https://github.com/RTradeLtd/go-datastores/issues/15
* Enable index creation for postgres as an optional functionality

### CODE REVIEW

File: sql/README.md [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+  * Upstream doesn't specifically test the postgres implementation which is likely to introduce bugs.
+* Use the `uber-go/multierr` package to combine errors where appropriatey
+    * In transaction mode if any errors happen, the error returned from the rollback was not returned, instead we combine the error for better visibility
+
+# Usage (Postgres)
+
+## Easy (Automatic)
+
+
+The `postgres` datastore has some helper utility to enable automatic table and index creation, as well as the ability to recreate the table by dropping it, and creating it. The following will make it easy to do this:
```
Comment (xiegeo) [inline sql/README.md]: Does this need `RunMigrations` set to true?

File: sql/README.md [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+)
+opts := &pgds.Options{
+	Host:                  "127.0.0.1",
+	Port:                  "5432",
+	User:                  "postgres",
+	Database:              "datastores",
+	Password:              "password123",
+    SSLMode:               "disable",
+    Table:                  "blocks",
+    // this will dropthe existing table
```
Comment (xiegeo) [inline sql/README.md]: dropthe

File: sql/postgres/postgres.go [hunk 3]
```diff
@@ -10,17 +10,28 @@ import (
...
-	Database       string
-	Table          string
-	SSLMode        string
-	RunMigrations  bool
-	RecreateTables bool
+	// must be set to "warning is accepted"
+	// this will prevent anyone from accidentally nuking their database
+	// the act of a specific string ensures that its more difficult
+	// to take this action
+	AcceptRecreateWarning string
```
Comment (xiegeo) [inline sql/postgres/postgres.go]: Comment for `AcceptRecreateWarning` does not match value for `AcceptTableRecreationWarning`

Also, one name has the word table in it while the other does not, I think this is confusing.

**Review by xiegeo — CHANGES_REQUESTED** [review]

File: sql/README.md (hunk 1 above)
Comment (bonedaddy) [reply sql/README.md]: yes it does, and the code wasn't check for that. good catch

File: sql/README.md (hunk 2 above)
Comment (bonedaddy) [reply sql/README.md]: good catch

File: sql/postgres/postgres.go (hunk 3 above)
Comment (bonedaddy) [reply sql/postgres/postgres.go]: good idea, i cleaned up the naming abit

**Review by xiegeo — APPROVED** [review]
Right now, the docs aren't very clear on when existing tables are deleted. Does it happen every time when the option is set, or only if the old table is not structured correctly?

**Comment by bonedaddy** [comment]
Good point I will update the docs. The tables are only deleted when `RecreateTables` is set, and the warning is accepted

**Comment by xiegeo** [comment]
> 
> 
> Good point I will update the docs. The tables are only deleted when `RecreateTables` is set, and the warning is accepted

So if the option is set and waring accepted, then the tables are always recreated on every program start?

**Comment by bonedaddy** [comment]
> So if the option is set and waring accepted, then the tables are always recreated on every program start?

Yes

**Pull request merged**